// react-native-graph/lib/module/SelectionDot.js
import React, { useCallback } from 'react';
import { runOnJS, useAnimatedReaction, useSharedValue, withSpring, useDerivedValue } from 'react-native-reanimated';
import { Circle, Group, Shadow } from '@shopify/react-native-skia';
export const CIRCLE_RADIUS = 5;
export const CIRCLE_RADIUS_MULTIPLIER = 6;
export function SelectionDot(_ref) {
  let {
    isActive,
    color,
    circleX,
    circleY
  } = _ref;
  const circleRadius = useSharedValue(0);
  const circleStrokeRadius = useDerivedValue(() => circleRadius.value * CIRCLE_RADIUS_MULTIPLIER, [circleRadius]);
  const setIsActive = useCallback(active => {
    circleRadius.value = withSpring(active ? CIRCLE_RADIUS : 0, {
      mass: 1,
      stiffness: 1000,
      damping: 50,
      velocity: 0
    });
  }, [circleRadius]);
  useAnimatedReaction(() => isActive.value, active => {
    runOnJS(setIsActive)(active);
  }, [isActive, setIsActive]);
  return /*#__PURE__*/React.createElement(Group, null, /*#__PURE__*/React.createElement(Circle, {
    opacity: 0.05,
    cx: circleX,
    cy: circleY,
    r: circleStrokeRadius,
    color: "#333333"
  }), /*#__PURE__*/React.createElement(Circle, {
    cx: circleX,
    cy: circleY,
    r: circleRadius,
    color: color
  }, /*#__PURE__*/React.createElement(Shadow, {
    dx: 0,
    dy: 0,
    color: "rgba(0,0,0,0.5)",
    blur: 4
  })));
}
